sale_product_set/wizard/product_set_add.py

```diff
@@ -0,0 +1,128 @@
+                # Fallback to all lines from current set
+                rec.product_set_line_ids = rec.product_set_id.set_line_ids
+
+    def _check_partner(self):
+        """Validate order partner against product set's partner if any."""
+        if not self.product_set_id.partner_id or self.env.context.get(
+            "product_set_add_skip_validation"
+        ):
+            return
+
+        allowed_partners = self._allowed_order_partners()
+        if self.order_id.partner_id not in allowed_partners:
+            raise exceptions.ValidationError(
+                _(
+                    "You can use a sale order assigned "
+                    "only to following partner(s): {}"
+                ).format(", ".join(allowed_partners.mapped("name")))
+            )
+
+    def _allowed_order_partners(self):
+        """Product sets' partners allowed for current sale order."""
+        partner_ids = self.env.context.get("allowed_order_partner_ids")
+        if partner_ids:
+            return self.env["res.partner"].browse(partner_ids)
+        return self.product_set_id.partner_id
+
+    def add_set(self):
+        """Add product set, multiplied by quantity in sale order line"""
+        self._check_partner()
+        order_lines = self._prepare_order_lines()
+        if order_lines:
+            self.order_id.write({"order_line": order_lines})
+        return order_lines
+
+    def _prepare_order_lines(self):
+        max_sequence = self._get_max_sequence()
+        order_lines = []
+        for seq, set_line in enumerate(self._get_lines(), start=1):
+            values = self.prepare_sale_order_line_data(set_line)
+            # When we play with sequence widget on a set of product,
+            # it's possible to have a negative sequence.
+            # In this case, the line is not added at the correct place.
+            # So we have to force it with the order of the line.
+            values.update({"sequence": max_sequence + seq})
+            order_lines.append((0, 0, values))
+        return order_lines
+
+    def _get_max_sequence(self):
+        max_sequence = 0
+        if self.order_id.order_line:
+            max_sequence = max(line.sequence for line in self.order_id.order_line)
+        return max_sequence
+
+    def _get_lines(self):
+        # hook here to take control on used lines
+        so_product_ids = self.order_id.order_line.mapped("product_id").ids
+        for set_line in self.product_set_line_ids:
+            if self.skip_existing_products and set_line.product_id.id in so_product_ids:
+                continue
+            yield set_line
+
+    def prepare_sale_order_line_data(self, set_line, max_sequence=0):
+        self.ensure_one()
+        line_values = set_line.prepare_sale_order_line_values(
+            self.order_id, self.quantity, max_sequence=max_sequence
+        )
+        sol_model = self.env["sale.order.line"]
+        line_values.update(sol_model.play_onchanges(line_values, line_values.keys()))
```

Comment:
@angelmoya We should remove the call to the play_onchanges method since fields on sale.order.line are no more initialized by onchange methods but by computed readonly methods... With this change the dependency on the `onchange_helper` can be removed.